### ISSUE TITLE
min/max aggregations have worse cpu performance with annotation functions

### ISSUE THREAD

**Issue by hackeryang** [trinodb/trino#15516]
We found that when a simple sql used min()/max() functions in our production environment, the machine CPU load displayed in Grafana became very high.  

So we checked the source codes, and found that the annotation function features are introduced in trino 383: https://github.com/trinodb/trino/pull/11477  

Then we made a test, here are the steps:    
(1) make a tpch hive table `lineitem` with about 20GB data, the table schema is like this:  
```
show create table test.hivelineitem;
             Create Table              
---------------------------------------
 CREATE TABLE hive.test.hivelineitem ( 
    orderkey bigint,                   
    partkey bigint,                    
    suppkey bigint,                    
    linenumber integer,                
    quantity double,                   
    extendedprice double,              
    discount double,                   
    tax double,                        
    returnflag varchar(1),             
    linestatus varchar(1),             
    shipdate date,                     
    commitdate date,                   
    receiptdate date,                  
    shipinstruct varchar(25),          
    shipmode varchar(10),              
    comment varchar(44)                
 )                                     
 WITH (                                
    format = 'ORC'                     
 )
```

(2) The test SQL are simple, they are `explain analyze select min(extendedprice) from test.hivelineitem;` and `explain analyze select max(extendedprice) from test.hivelineitem;`. 
(3) We compared the CPU time in query plans between trino 380 and 392, and the results looks like this:  
a. min() in 380  
![b173220f-3983-420b-bd0f-871ce85a3c06](https://user-images.githubusercontent.com/26461591/209256454-b7671603-d21e-4b45-807d-cf2c6cdfdc23.jpeg)  
b. min() in 392
![8a2f41bf-9bc0-4bd3-aef6-b6c908a1ad6f](https://user-images.githubusercontent.com/26461591/209256564-f3058f9a-771e-4500-81dc-0b35fe5393d2.jpeg)  
c. max() in 380  
![f3c8418b-9072-4081-8d6e-22ffa2ed6ddc](https://user-images.githubusercontent.com/26461591/209256635-e9f3edd9-e059-4626-9302-161eb0450893.jpeg)  
d. max() in 392  
![b7d1cede-02e2-4796-8737-65dc4da8a1d3](https://user-images.githubusercontent.com/26461591/209256680-095d48a2-d322-4fa9-bbb9-aa8990503ed0.jpeg)  

So, according to the result pictures above, you can see that the CPU time used in PARTIAL AGGREGATION from Fragment 2 with version 392 is much higher than that in 380.

**Comment by hackeryang** [comment]
Hi @dain, can you please take a look when you have time~thank you  
And we found that when we revert the patch `bdbfc012` temporarily in version 392, the cpu performance became better, here is the result after reversion:  
![9af756ca-4335-4cdd-87a7-714378c29bfd](https://user-images.githubusercontent.com/26461591/209257329-3f178677-3614-4558-aad2-346682124528.jpeg)  
![9b6ad54f-a7b1-47ed-acc3-40f768b23619](https://user-images.githubusercontent.com/26461591/209257388-4bc20556-088a-4cf1-b77f-3b1757ee9b22.jpeg)

**Comment by raunaqmorarka** [comment]
fyi @kabunchi @lukasz-stec @sopel39

**Comment by sopel39** [comment]
Cc @dain

**Comment by lukasz-stec** [comment]
@hackeryang did you run the queries on the freshly started cluster or was there some warmup (some queries were run before the min/max)?
what hardware are you using? arm (e.g. graviton) or x86?

**Comment by hackeryang** [comment]
> @hackeryang did you run the queries on the freshly started cluster or was there some warmup (some queries were run before the min/max)? what hardware are you using? arm (e.g. graviton) or x86?

@lukasz-stec  The high cpu phenomenon in the Grafana of our production environment could reproduce 100% every day when  a min/max sql was running before, so we made that test above.  
We ran that test on a idle cluster without any other queries running, which has 3 nodes with x86 CPU, the model is: `Intel(R) Xeon(R) CPU E5-2682 v4 @ 2.50GHz(4 core with Hyper-threading)`, and we didn't run any warm up queries.  

Our `jvm.config` about the test cluster is like this, we used Azul zulu jdk 11.0.15:  
```
-server
-Xmx22G
-XX:-UseBiasedLocking
-XX:+UseG1GC
-XX:G1HeapRegionSize=32M
-XX:+ExplicitGCInvokesConcurrent
-XX:+UseGCOverheadLimit
-XX:+ExitOnOutOfMemoryError
-XX:ReservedCodeCacheSize=512M
-Djdk.attach.allowAttachSelf=true
-Djdk.nio.maxCachedBufferSize=2000000
-XX:PerMethodRecompilationCutoff=10000
-XX:PerBytecodeRecompilationCutoff=10000
-XX:-OmitStackTraceInFastThrow
-agentlib:jvmkill
--add-exports=java.base/sun.net.util=ALL-UNNAMED
```

**Comment by lukasz-stec** [comment]
> we used Azul zulu jdk 11.0.15:

Are you sure you are running trino 392? Since 390, trino requires java `17.0.3` or higher.

**Comment by hackeryang** [comment]
> > we used Azul zulu jdk 11.0.15:
> 
> Are you sure you are running trino 392? Since 390, trino requires java `17.0.3` or higher.

@lukasz-stec Yes, we reverted the jdk17 patch `d78c38b8` in our internal trino for some performance and compatibility issue met in the production environment.

**Comment by sopel39** [comment]
> @lukasz-stec Yes, we reverted the jdk17 patch d78c38b8 in our internal trino for some performance and compatibility issue met in the production environment.

Would it be possible to test if regression still exists with jdk17?

**Comment by hackeryang** [comment]
> > @lukasz-stec Yes, we reverted the jdk17 patch [d78c38b](https://github.com/trinodb/trino/commit/d78c38b8589fb8756e23a10af105298c06c1cb3f) in our internal trino for some performance and compatibility issue met in the production environment.
> 
> Would it be possible to test if regression still exists with jdk17?

@sopel39 Sure, we launched the trino 392 with `zulu jdk 17.0.4`, and the test resuts are below:    
(1) min()  with annotation function(first time was about 50s, the second and third time were about 5s):
![image](https://user-images.githubusercontent.com/26461591/210724028-7dcf708a-adbb-4159-9c70-c124df09be02.png)  
(2) min() without annotation function:  
![image](https://user-images.githubusercontent.com/26461591/210751491-793829f4-825d-40a9-a95e-10108a101711.png)  
(3) max() with annotation function:  
![image](https://user-images.githubusercontent.com/26461591/210726519-7a9bda61-37a1-49aa-a054-95757bfbd729.png)  
(4) max() without annotation function:  
![image](https://user-images.githubusercontent.com/26461591/210751983-c5ae373d-5d9d-4eb6-801d-04e43e639844.png)  

So if we use jdk 17, the aggregation performance with annotation functions will be better than without annotation functions ,and be much better than jdk 11, but if we use jdk 11, the performance will be better if we don't use annotation functions.

So the regression range is from trino 383 to 390.

**Comment by sopel39** [comment]
Right. In newest versions we disallow java <17, so I think we can actually close this issue.

**Comment by hackeryang** [comment]
> Right. In newest versions we disallow java <17, so I think we can actually close this issue.

OK, thank you for your advice~Yes i have no other questions, we will change to jdk 17 in the newer version.
